File: location/src/main/java/akhoi/libs/mlct/tools/java/ByteConcatReader.java
package akhoi.libs.mlct.tools.java;

public class ByteConcatReader {
    private final byte[] content;
    private int byteIndex;
    private int position;

    public ByteConcatReader(byte[] content) {
        this.content = content;
    }

    public int readInt(int size) {
        int remaining = Math.max(Math.min(size, 32), 0);
        if (remaining == 0) {
            return 0;
        }
        int remainder = 8 - position;
        int readSize = Math.min(remaining, remainder);
        int result = readByte(0, readSize, remainder);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }
        readSize = Math.min(remaining, 8);
        result = readByte(result, readSize, 8);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }
        readSize = Math.min(remaining, 8);
        result = readByte(result, readSize, 8);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }
        readSize = Math.min(remaining, 8);
        result = readByte(result, readSize, 8);
        remaining -= readSize;
        if (remaining <= 0) {
            return result;
        }
        readSize = Math.min(remaining, 8);
        return readByte(result, readSize, 8);
    }

    public long readLong(int size) {
        if (size <= 32) {
            long value = readInt(size);
            return value & ((1L << size) - 1);
        } else {
            int msBitsCount = size - 32;
            long msBits = readInt(msBitsCount) & ((1L << msBitsCount) - 1);
            long lsBits = readInt(32) & 0xFFFFFFFFL;
            return (msBits << 32) | lsBits;
        }
    }

    public double readDouble(int size) {
        long longValue = readLong(size);
        return Double.longBitsToDouble(longValue);
    }

    public void reset() {
        position = 0;
    }

    public void skip(int count) {
        position += count;
    }

    private int readByte(int value, int size, int remainder) {
        if (byteIndex >= content.length) {
            return value;
        }
        int read = ((1 << size) - 1) & ((content[byteIndex] & 0xFF) >>> (remainder - size));
        int nextBytePosition = position + size;
        byteIndex += nextBytePosition >>> 3;
        position = nextBytePosition & 7;
        return (value << size) | read;
    }
}

File: location/src/main/java/akhoi/libs/mlct/tools/ByteConcatReader.kt
package akhoi.libs.mlct.tools

import kotlin.math.max
import kotlin.math.min

class ByteConcatReader(private val content: ByteArray) {
    private var byte: Int = 0
    private var position: Int = 0

    fun readInt(size: Int): Int {
        var remaining = max(min(size, 32), 0)
        if (remaining == 0) return 0

        val remainder = 8 - position
        var readSize = min(remaining, remainder)
        var result = readByte(0, readSize, remainder)
        remaining -= readSize
        if (remaining <= 0) return result

        readSize = min(remaining, 8)
        result = readByte(result, readSize, 8)
        remaining -= readSize
        if (remaining <= 0) return result

        readSize = min(remaining, 8)
        result = readByte(result, readSize, 8)
        remaining -= readSize
        if (remaining <= 0) return result

        readSize = min(remaining, 8)
        result = readByte(result, readSize, 8)
        remaining -= readSize
        if (remaining <= 0) return result

        readSize = min(remaining, 8)
        result = readByte(result, readSize, 8)

        return result
    }

    fun readLong(size: Int): Long {
        val value: Long
        if (size <= 32) {
            value = readInt(size).toLong() and (1L shl size) - 1
        } else {
            val msBitsCount = size - 32
            val msBits = readInt(msBitsCount).toLong() and (1L shl msBitsCount) - 1
            val lsBits = readInt(32).toLong() and 0xFFFFFFFF
            value = msBits shl 32 or lsBits
        }
        return value
    }

    fun readDouble(size: Int): Double {
        val longValue = readLong(size)
        return java.lang.Double.longBitsToDouble(longValue)
    }

    fun reset() {
        position = 0
    }

    fun skip(count: Int) {
        position += count
    }

    private inline fun readByte(value: Int, size: Int, remainder: Int): Int {
        if (byte >= content.size) return value

        val read = ((1 shl size) - 1) and (content[byte].toInt() ushr (remainder - size))
        val nextBytePosition = position + size
        byte += nextBytePosition shr 3
        position = nextBytePosition and 7
        return (value shl size) or read
    }
}